File: scripts/serve.ts
import { join, resolve } from 'path';
import { statSync, readdirSync } from 'fs';
import chalk from 'chalk';

const port = 4213;
const rootDir = resolve(import.meta.dir, '..');
const distDir = join(rootDir, 'dist');
const demoDir = join(rootDir, 'demo');

// 检查dist目录是否存在
try {
    const distStat = statSync(distDir);
    if (!distStat.isDirectory()) {
        console.error(chalk.red('Error: dist is not a directory'));
        process.exit(1);
    }
} catch (err) {
    console.error(chalk.red('Error: dist directory does not exist. Please run `bun run build` first.'));
    process.exit(1);
}

console.log(chalk.cyan('Starting development server...'));
console.log(chalk.green('Server directories:'));
console.log(chalk.green(`- dist: ${distDir}`));
console.log(chalk.green(`- demo: ${demoDir}`));

// 构建可用字体集列表
const fontSets = [];
try {
    // 查找所有HTML演示文件
    const distFiles = readdirSync(distDir);
    for (const file of distFiles) {
        if (file.endsWith('.html')) {
            const fontCode = file.replace('.html', '');
            fontSets.push({
                code:    fontCode,
                demoUrl: `/fonts/${fontCode}`,
            });
        }
    }

    // 检查子目录
    for (const item of readdirSync(distDir)) {
        const itemPath = join(distDir, item);
        const itemStat = statSync(itemPath);
        if (itemStat.isDirectory()) {
            fontSets.push({
                code:    item,
                demoUrl: `/fonts/${item}`,
            });
        }
    }
} catch (err) {
    console.log(chalk.yellow('Warning: Failed to build font sets list.'), err);
}

// 生成字体索引页面
function generateFontIndexHtml() {
    let fontListHtml = '';

    for (const font of fontSets) {
        fontListHtml += `
            <li>
                <a href="${font.demoUrl}" class="font-link">
                    <span class="font-code">${font.code}</span>
                </a>
            </li>`;
    }

    return `<!DOCTYPE html>
    <html lang="en">
    <head>
        <meta charset="UTF-8">
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        <title>Chromana Font Collection</title>
        <style>
            body {
                font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, 'Open Sans', 'Helvetica Neue', sans-serif;
                margin: 0;
                padding: 40px 20px;
                background-color: #f5f5f5;
                color: #333;
            }
            h1 {
                text-align: center;
                margin-bottom: 40px;
            }
            .container {
                max-width: 800px;
                margin: 0 auto;
            }
            .font-list {
                list-style: none;
                padding: 0;
                display: grid;
                grid-template-columns: repeat(auto-fill, minmax(200px, 1fr));
                gap: 16px;
            }
            .font-link {
                display: block;
                text-decoration: none;
                background-color: white;
                padding: 20px;
                border-radius: 8px;
                box-shadow: 0 2px 4px rgba(0,0,0,0.1);
                text-align: center;
                transition: all 0.3s ease;
            }
            .font-link:hover {
                transform: translateY(-3px);
                box-shadow: 0 4px 8px rgba(0,0,0,0.15);
            }
            .font-code {
                font-size: 18px;
                font-weight: 500;
                color: #2c5282;
            }
            .note {
                margin-top: 40px;
                padding: 15px;
                background-color: #e6fffa;
                border-left: 4px solid #38b2ac;
                border-radius: 4px;
            }
        </style>
    </head>
    <body>
        <div class="container">
            <h1>Chromana Font Collection</h1>

            <ul class="font-list">
                ${fontListHtml}
            </ul>

            <div class="note">
                <p><strong>注意：</strong> 点击上方链接查看各个字体的演示页面。每个字体都提供了TTF、WOFF和WOFF2格式。</p>
            </div>
        </div>
    </body>
    </html>`;
}

Bun.serve({
    port,
    async fetch(req) {
        const url = new URL(req.url);
        const path = url.pathname;

        // 默认路径
        if (path === '/') {
            // 如果是根路径，首先尝试提供演示目录的index.html
            try {
                const demoIndex = join(demoDir, 'index.html');
                const stat = statSync(demoIndex);
                if (stat.isFile()) {
                    const file = Bun.file(demoIndex);
                    return new Response(file);
                }
            } catch (err) {
                // 如果demo目录的index.html不存在，提供字体索引页面
                return new Response(generateFontIndexHtml(), {
                    headers: { 'Content-Type': 'text/html' },
                });
            }
        }

        // 字体演示路径
        if (path.startsWith('/fonts/')) {
            const fontCode = path.slice(7); // 去掉'/fonts/'

            // 尝试找到相应的HTML文件
            let htmlPath;

            if (fontCode.includes('/')) {
                // 如果是子目录路径，如/fonts/magic/W.svg
                const relativePath = fontCode;
                htmlPath = join(distDir, relativePath);
            } else {
                // 如果是字体代码，如/fonts/magic
                htmlPath = join(distDir, `${fontCode}.html`);
                if (!statSync(htmlPath, { throwIfNoEntry: false })) {
                    // 如果根目录没有对应的HTML，检查是否是子目录
                    const fontDir = join(distDir, fontCode);
                    if (statSync(fontDir, { throwIfNoEntry: false })?.isDirectory()) {
                        // 如果是子目录，查找子目录中的HTML文件
                        try {
                            const dirFiles = readdirSync(fontDir);
                            const htmlFile = dirFiles.find(f => f.endsWith('.html'));
                            if (htmlFile) {
                                htmlPath = join(fontDir, htmlFile);
                            }
                        } catch (err) {
                            // 忽略错误
                        }
                    }
                }
            }

            if (htmlPath && statSync(htmlPath, { throwIfNoEntry: false })?.isFile()) {
                const file = Bun.file(htmlPath);
                return new Response(file);
            }
        }

        let filePath;

        // 如果路径以/dist开头，则从dist目录中提供文件
        if (path.startsWith('/dist/')) {
            filePath = join(rootDir, path);
        } else if (path.startsWith('/fonts/')) {
            // 字体资源路径 /fonts/magic.ttf -> /dist/magic.ttf
            const fontPath = path.slice(7); // 去掉'/fonts/'
            filePath = join(distDir, fontPath);
        } else {
            // 否则从demo目录提供文件
            filePath = join(demoDir, path);
        }

        try {
            const stat = statSync(filePath);
            if (stat.isFile()) {
                const file = Bun.file(filePath);
                return new Response(file);
            }
        } catch (err) {
            // 文件不存在
            console.log(chalk.yellow(`404: ${path}`));
            return new Response('Not Found', { status: 404 });
        }

        return new Response('Not Found', { status: 404 });
    },
});

console.log(chalk.green(`Server running at http://localhost:${port}`));
console.log(chalk.green('Press Ctrl+C to stop'));

File: scripts/serve.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import os
import sys
import http.server
import socketserver
from pathlib import Path

# 项目根目录
PROJECT_ROOT = Path(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
DIST_DIR = PROJECT_ROOT / "dist"
DEMO_DIR = PROJECT_ROOT / "demo"

# 默认端口
PORT = 4213

# 检查dist目录是否存在
if not DIST_DIR.exists() or not DIST_DIR.is_dir():
    print("Error: dist directory does not exist. Please run `python scripts/build.py` first.")
    sys.exit(1)

print("Starting development server...")
print(f"Server directories:")
print(f"- dist: {DIST_DIR}")
print(f"- demo: {DEMO_DIR}")

# 自定义请求处理器，用于提供来自不同目录的文件
class ChromanaRequestHandler(http.server.SimpleHTTPRequestHandler):
    def translate_path(self, path):
        # 默认路径处理
        path = super().translate_path(path)

        # 处理根路径 - 显示索引页面
        rel_path = os.path.relpath(path, os.getcwd())
        if rel_path == '.' or rel_path == './':
            # 提供演示根目录的索引
            return str(DEMO_DIR / "index.html")

        # 检查demo目录
        demo_path = DEMO_DIR / os.path.relpath(path, os.getcwd())
        if demo_path.exists():
            return str(demo_path)

        # 检查dist目录
        dist_path = DIST_DIR / os.path.relpath(path, os.getcwd())
        if dist_path.exists():
            return str(dist_path)

        return path

    def log_message(self, format, *args):
        # 彩色日志输出
        if args[1].startswith('2'): # 2xx状态码为成功
            status = f"\033[92m{args[1]}\033[0m" # 绿色
        elif args[1].startswith('3'): # 3xx状态码为重定向
            status = f"\033[94m{args[1]}\033[0m" # 蓝色
        elif args[1].startswith('4') or args[1].startswith('5'): # 4xx, 5xx为错误
            status = f"\033[91m{args[1]}\033[0m" # 红色
        else:
            status = args[1]

        print(f"{self.address_string()} - {status} {args[0] % args[1:]}")

# 启动服务器
handler = ChromanaRequestHandler
with socketserver.TCPServer(("", PORT), handler) as httpd:
    print(f"\nServer running at http://localhost:{PORT}/")
    print("Press Ctrl+C to stop")
    try:
        httpd.serve_forever()
    except KeyboardInterrupt:
        print("\nShutting down server...")
        httpd.server_close()
